jxl/src/decode.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    bit_reader::BitReader,
+    enc::{ImageData, ImageFrame},
+    error::Error,
+    frame::{DecoderState, Frame, Section},
+    headers::{FileHeader, JxlHeader},
+    icc::read_icc,
+};
+
+use std::marker::PhantomData;
+
+#[allow(clippy::type_complexity)]
+pub struct DecodeOptions<'a> {
+    pub xyb_output_linear: bool,
+    #[cfg(test)]
+    pub frame_callback: Option<&'a mut dyn FnMut(&Frame) -> Result<(), Error>>,
```

Comment:
Why is this not generic on an owned callback type?
i.e.
```
pub struct DecodeOptions<F: FnMut(&Frame) -> Result<(), Error>> {
...
}
```

For that matter, why is it `#[cfg(test)]`?

It should be relatively easy to have something like
```
impl DecodeOptions<fn(&Frame) -> Result<(), Error>> {
  pub fn new() {
     xyb_output_linear: true,
     frame_callback: None,
  }
}
```
and then have `new_for_testing` on generic `F` which takes the callback.

Reply:
Removed cfg(test), but kept the lifetime, because that is needed by the tests.